docs: add glossary, `add` explainer, and content routing FAQ (#233)
filecoin-project/filecoin-pin: documentation/behind-the-scenes-of-adding-a-file.md

## Purpose

The steps outlined below are taken to "add a file with Filecoin Pin".  This document is intended to provide more info about what happens "behind the scenes" as it uses underlying libraries like [`synapse`](glossary.md#synapse) and the [Filecoin Onchain Cloud](glossary.md#filecoin-onchain-cloud) offering.

## Diagram

_Key_ 
Blue | Orange | FP | SP
-- | -- | -- | --

> non-blockchain step | blockchain step | Filecoin Pin | Storage Provider

Reviewer: I seem to remember some convo when I first joined where we were attempting to change "storage provider" to "service provider".. not sure if that's still the case.

```suggestion
Blue | Orange | FP | SP
-- | -- | -- | --
non-blockchain step | blockchain step | Filecoin Pin | Service Provider
```

Author: Yeah, you're right.  I don't know why I did Storage Provider throughout this document.  I changing to Service Provider.

```mermaid
graph TD

    %% Node Definitions
    %% Non-blockchain nodes

>     Start([User: Select File to Add])

Reviewer 2: is this just for the website? because this could be "Select File or Directory to Add"

    FilecoinPinAddDone --> RetrieveData

    %% Blockchain flow
    ConnectWallet --> SetupPay
    SetupPay --> IdentifyDataSet
    IdentifyDataSet --> CreateDataSet

    %% Convergence
    UploadCAR --> CreateDataSet
    CreateDataSet --> ConfirmBlockchainTx
    ConfirmBlockchainTx --> FilecoinPinAddDone
    FilecoinPinAddDone --> ProveData

    %% Styling
    classDef nonBlockchain fill:#e1f5ff,stroke:#0288d1,stroke-width:2px,color:#000
    classDef blockchain fill:#fff3e0,stroke:#f57c00,stroke-width:2px,color:#000

    class Start,CreateCAR,UploadCAR,IndexCAR,AdvertiseCAR,AwaitIPNIIndexing,RetrieveData nonBlockchain
    class ConnectWallet,SetupPay,IdentifyDataSet,CreateDataSet,ConfirmBlockchainTx,ProveData blockchain
```

## Steps without Blockchain Interactions

These are the set of steps that are done client side (i.e., where the Filecoin Pin code is running) and with a [Storage Provider](glossary.md#storage-provider) that don't involve the Filecoin blockchain.  These steps in isolation though don't yield a committed cryptographic proof of the data being possessed by and retrievable from an SP, but they are necessary preconditions.

### Create CAR

*What/why:*

> The provided file needs to be turned into a merkle DAG and have the DAG's blocks shipped off.  [CAR](glossary.md#car) is a common container format for transporting blocks in the IPFS ecosystem and is used with Filecoin Pin.  Filecoin Pin DAGifies the content so that it knows the CIDs of the blocks so it doesn't need to trust the [Storage Provider](glossary.md#storage-provider) (SP).

Reviewer 2: ```suggestion
The provided file needs to be turned into a merkle DAG and have the DAG's blocks shipped off.  [CAR](glossary.md#car) is a common container format for transporting blocks in the IPFS ecosystem and is used with Filecoin Pin.  Storage providers store and prove contiguous sequence of bytes, but generating IPFS compatible data from files and directories creates potentially very many small "blocks" of data, which we pack into a single CAR container for storage providers to store and prove. The process of packing in IPFS form, or "DAGifying" the file data, allows us to reference and verify smaller units of our content, and gives us the ability to interact _trustlessly_ with storage providers to serve and retrieve our file data.
```

or something like that

Implementation notes:

> - The CAR file is created using Helia for UnixFS DAG creation.

Reviewer: idk if we care about diving into the details, but we do have some car creation nuance

Author: I'll wait to add more details until it gets asked about more.

The SP produces a new or updated advertisement chain.  By the end, IPNI indexers should have additional provider records for the advertised CIDs.

*Expected duration:*

Local indexing of the CAR is quick as the CAR already contains a list of CIDs and their offsets, which is verified and reused.  Creating/updating an advertisement chain and announcing it to IPNI indexers is also quick.  There is a delay in an IPNI indexer on the order of seconds for coming to grab the advertisements plus some ingestion delay on the IPNI indexer side.

## Blockchain related steps

Below are the set of steps that are particularly unique from traditional IPFS usage as they involve authorization, payment, and cryptographic proofs.

### Connect Wallet

*What/why:*

Filecoin Pin needs to interface with the Filecoin blockchain to authorize and send payment to [storage providers](glossary.md#storage-provider) (SPs) for their work of storing and proving possession of data.  This requires having a secret key to sign messages sent to the blockchain.

> Currently `filecoin-pin` expects to be explicitly passed a private key via environment variable or command line argument.  [filecoin-pin-website](glossary.md#filecoin-pin-website) as [pin.filecoin.cloud](http://pin.filecoin.cloud) uses a global [session key](glossary.md#session-key) which can be embedded into source code since it scopes down the set of actions that can be performed.

Reviewer 2: ```suggestion
Currently `filecoin-pin` expects to be explicitly passed a private key via environment variable or command line argument.  [filecoin-pin-website](glossary.md#filecoin-pin-website) as [pin.filecoin.cloud](http://pin.filecoin.cloud) uses a global [session key](glossary.md#session-key) which can be embedded into source code since it scopes down the set of actions that can be performed and alleviating the need for a user to provide a wallet to perform operations.
```

*Outputs:*

A record onchain denoting the data that needs to periodically be proven to be in the possession of the Data Set’s SP.

*Expected duration:*

As a single transaction, this takes ~30 seconds to be confirmed onchain.  

## Content Routing FAQ

### Will indexed CIDs from Calibration be mixed with CIDs from Mainnet?

Yes.  [IPNI](glossary.md#ipni) indexers are not chain aware.  They key on the CID and will point to whatever providers have "recently" advertised the CID.  This means that if a given piece is created with a [Calibration](glossary.md#calibration-network) SP and also with a Mainnet SP, the CIDs will list both SPs as providers.

### What happens when a piece is deleted?

> When an SP is instructed to delete a [piece](glossary.md#piece), it announces a new advertisement to [IPNI](glossary.md#ipni) that includes the removal of the CIDs within the piece.  This update to IPNI goes through the normal IPNI flow of receiving advertisement announcements and then asynchronously fetching the advertisements from the provider.  As a result, deleted pieces should take seconds to low minutes for IPNI index state to be updated.

Author: I think we leave the full delete story for another day given this doc PR has already taken awhile to deliver and we don't have delete functionality in filecoin-pin or filecoin-pin-website yet.
